precompiles/ics20/approve_common.go

```diff
@@ -0,0 +1,267 @@
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	sourcePort, sourceChannel, denom string,
+	amount *big.Int,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
+	msgAuthz, expiration, err := authorization.CheckAuthzExists(ctx, authzKeeper, grantee, origin, TransferMsgURL)
+	if err != nil {
+		return fmt.Errorf(authorization.ErrAuthzDoesNotExistOrExpired, grantee, origin)
+	}
+
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
+	transferAuthz, ok := msgAuthz.(*transfertypes.TransferAuthorization)
+	if !ok {
+		return authz.ErrUnknownAuthorizationType
+	}
+
+	// Check if the allocations matches the arguments provided and returns the index of the allocation and coin found
+	spendLimit, allocationIdx, err := checkAllocationExists(transferAuthz.Allocations, sourcePort, sourceChannel, denom)
+	if err != nil {
+		return err
+	}
+
+	allowance := sdk.NewIntFromBigInt(amount)
+	if _, overflow := cmn.SafeAdd(spendLimit.Amount, allowance); overflow {
+		return errors.New(cmn.ErrIntegerOverflow)
+	}
+
+	allowanceCoin := sdk.Coin{Denom: denom, Amount: allowance}
+
+	transferAuthz.Allocations[allocationIdx].SpendLimit = transferAuthz.Allocations[allocationIdx].SpendLimit.Add(allowanceCoin)
+
+	if err = authzKeeper.SaveGrant(ctx, grantee.Bytes(), origin.Bytes(), transferAuthz, expiration); err != nil {
+		return err
+	}
+
+	allocations := convertToAllocation(transferAuthz.Allocations)
+	// Emit the IBC transfer authorization event
+	return authorization.EmitIBCTransferAuthorizationEvent(
+		event,
+		ctx,
+		stateDB,
+		precompileAddr,
+		grantee,
+		origin,
+		allocations,
+	)
+}
+
+// DecreaseAllowance implements the ICS20 Authorization decrease allowance transactions.
+func DecreaseAllowance(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	sourcePort, sourceChannel, denom string,
+	amount *big.Int,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
+	msgAuthz, expiration, err := authorization.CheckAuthzExists(ctx, authzKeeper, grantee, origin, TransferMsgURL)
+	if err != nil {
+		return fmt.Errorf(authorization.ErrAuthzDoesNotExistOrExpired, grantee, origin)
+	}
+
+	transferAuthz, ok := msgAuthz.(*transfertypes.TransferAuthorization)
+	if !ok {
+		return authz.ErrUnknownAuthorizationType
+	}
+
+	// Check if the allocations matches the arguments provided and returns the index of the allocation and spend limit found
+	spendLimit, allocationIdx, err := checkAllocationExists(transferAuthz.Allocations, sourcePort, sourceChannel, denom)
+	if err != nil {
+		return err
+	}
+
+	expense := sdk.NewIntFromBigInt(amount)
+	if spendLimit.Amount.LT(expense) {
+		return fmt.Errorf(cmn.ErrNegativeAmount)
```

Comment:
I would add to the error message something more informative, like "trying to decrease more than spend limit".